### ISSUE TITLE
[#3164] research_rl_ppo: PPO actor fige a l'init (bug total_loss) — verdict NO BEATS mesure une policy aleatoire

### ISSUE THREAD

**Issue by jsboige** [jsboige/CoursIA#3326]
## Contexte

Audit de fidelite **#3164** (lot `research_*`). Finding **REINVENTION involontaire** sur `MyIA.AI.Notebooks/QuantConnect/research/research_rl_ppo.ipynb` (auteur : myia-po-2024, cycle anterieur #1454/#1461).

## Bug (confirme par lecture directe cell-6, G.1)

Dans `PPOAgent.update()` (cell-6), le bloc finite-difference de l'actor est **incomplet** :

```python
total_loss = 0            # ligne 318 : init, JAMAIS reassigne dans la boucle actor
eps = 0.001
params = self.actor.get_all_params()
grad = np.zeros_like(params)
for i in range(0, len(params), ...):
    ...
    loss_plus = ...       # loss aux params perturbes (correct)
    grad[i] = (loss_plus - total_loss) / eps if total_loss != 0 else 0   # ligne 339
if np.linalg.norm(grad) > 0:        # toujours False (grad == 0 partout)
    self.actor.set_all_params(params + self.actor.lr * grad)            # JAMAIS appele
```

Le guard `if total_loss != 0 else 0` est le smoking gun : check defensif pour une valeur **jamais calculee**. `total_loss` devait etre la baseline loss `L(theta)` aux params actuels (point de reference de la finite-difference), mais elle reste a 0. Resultat : `grad` est identiquement nul, l'actor n'est **jamais** mis a jour pendant tout l'entrainement (200 episodes x 4 seeds). Le critic, lui, s'update (boucle TD separee, ~ligne 358).

## Preuve
- `grep -n "total_loss" research_rl_ppo.ipynb` : 1 init (ligne 318), 0 reassignation dans la boucle actor, 1 return (ligne 361, retourne 0).
- Lecture cell-6 : confirme l'absence de calcul de baseline loss.

## Consequence fidelite (#3164)

Le "PPO" entraine = **policy aleatoire figee a l'init**. Le verdict pedagogique `NO BEATS` (Sharpe mean -0.338 vs B&H 1.417, 0/4 seeds positifs, sigma edge -0.31) est **correct par accident** mais pour la **mauvaise raison** : il mesure une policy non-entrainee, pas PPO. La conclusion "PPO ne bat pas B&H sur SPY" est donc **invalide** — non-fidele au portage de Schulman et al. 2017, qui requiert une ascent effective sur L^CLIP.

Le notebook est **partiellement honnete** : il note "approximation grossiere du gradient" (cell-15) et "prototype pedagogique" (cell-16), suggere PyTorch pour RL-03+. Mais "approximation grossiere" sous-entend "apprend mal" — alors qu'en realite l'actor **n'apprend jamais**. La distinction est pedagogiquement cruciale.

## Fix propose (chirurgical, ~10 lignes, cell-6)

Calculer la baseline `total_loss` (loss aux params actuels) **avant** la boucle finite-diff, et retirer le guard cassant :

```python
# Baseline loss aux params actuels (point de reference finite-difference)
total_loss = 0
for s, a, old_p, adv_val in zip(states, actions, old_probs, adv):
    logits = self.actor.forward(s)
    probs = softmax(logits)
    new_p = max(probs[a], 1e-8)
    ratio = new_p / max(old_p, 1e-8)
    clipped = min(max(ratio, 1 - self.clip_eps), 1 + self.clip_eps)
    total_loss += min(ratio * adv_val, clipped * adv_val)
eps = 0.001
params = self.actor.get_all_params()
grad = np.zeros_like(params)
for i in range(0, len(params), max(1, len(params) // 50)):
    ...
    grad[i] = (loss_plus - total_loss) / eps   # guard retire
```

L'actor apprendra rellement (finite-diff bruite). Overhead negligeable (1 forward-loop/update vs 50 existantes). Re-execution 4 seeds x 200 episodes (numpy, ~minutes). Verdict honnete : probablement toujours NO BEATS (finite-diff PPO sur signal financier faible), mais pour la **bonne** raison cette fois.

## Conformite
- Anti-regression : le fix **restaure l'intention** (finite-diff policy gradient), ne remplace pas par sorry/stub. Conforme.
- C.2/H.3 : re-execution complete apres modif cellule code.

See #3164

**Comment by jsboige** [comment]
**Update — second bug découvert pendant le fix (cause racine de l'insensibilité)**

En implementant le fix du bug `total_loss`, un sanity-check a révélé que l'actor restait fige **meme apres** le calcul de la baseline. Diagnostic par lecture + test direct : il y a un **second bug complementaire** dans `SmallMLP.set_all_params()` (cell-6) :

```python
def set_all_params(self, flat):
    idx = 0
    for i, p in enumerate(self.params):
        size = p.size
        self.params[i] = flat[idx:idx+size].reshape(p.shape)   # rebind self.params[i]...
        idx += size
    # ...mais self.W1/b1/W2/b2 (lus par forward) ne sont PAS resyncs
```

`forward()` lit `self.W1` (l'attribut), pas `self.params[0]`. Apres `set_all_params`, `self.params[0]` pointe vers un nouvel array mais `self.W1` garde l'original. Preuve directe : `flat[0] += 999` puis `set_all_params(flat)` donne `self.params[0][0,0]=999.013` mais `self.W1[0,0]=0.0126` (inchange), et `self.params[0] is self.W1` → **False**.

**Consequence** : les perturbations finite-difference sont **invisibles** pour `forward()` → `loss_plus == total_loss` (les deux valent la meme chose car la forward ne change pas) → `grad == 0` → actor jamais mis a jour. Ce bug masquait le bug `total_loss` (et vice-versa) : corriger l'un sans l'autre laisse l'actor fige.

## Fix complet (2 corrections chirurgicales, cell-6)

1. **Calculer la baseline `total_loss`** (loss aux params actuels) avant la boucle finite-diff + retirer le guard `if total_loss != 0 else 0`. [deja decrit dans le body]
2. **Resync les attributs** dans `set_all_params` : `self.W1, self.b1, self.W2, self.b2 = self.params` apres la boucle.

## Verification (sanity-check pre-reexec)

Apres les 2 fixes, avec advantages non-uniformes (sinon la normalisation les annule) :
- `set_all_params(flat[0]+=999)` → `self.W1[0,0]=999.013`, `params[0] is W1=True` (resync OK)
- `actor.W1 max|delta|` apres 1 `update()` = **2.47e-04** → l'actor APPREND (vs 0.0 avant)

Re-execution Papermill complete (4 seeds x 200 episodes) en cours pour confirmer le verdict honnete post-fix.

See #3164
